ansibullbot/utils/git_tools.py

```diff
@@ -157,27 +157,12 @@ def manage_checkout(self):
 
     def get_files(self, force=False):
         '''Cache a list of filenames in the checkout'''
-        if self.isgit:
-            if not self._files or force:
-                cmd = f'cd {self.checkoutdir}; git ls-files'
-                logging.debug(cmd)
-                (rc, so, se) = run_command(cmd)
-                files = to_text(so).split('\n')
-                files = [x.strip() for x in files if x.strip()]
-                if self.context:
-                    self._files = [x for x in files if self.context in files]
-                self._files = files
-        else:
-            self._files = []
-            cmd = f'cd {self.checkoutdir}; find .'
-            logging.debug(cmd)
-            (rc, so, se) = run_command(cmd)
-            filepaths = to_text(so).split('\n')
-            for fp in filepaths:
-                if not fp.startswith('./'):
-                    continue
-                fp = fp.replace('./', '', 1)
-                self._files.append(fp)
+        if not self._files or force:
```

Comment:
I understand the desire to rewrite this method but what was the reason to do this besides the code being "not optimal"?

The new code, as opposed to the old one, returns also all files from the `.git` directory which I am not sure is desired, `git ls-files` takes care of filtering such files (but not limited to). So this should be taken into account.

Also, the new code seems to ignore `self.context` which appears to be fine as `self.context` is not set anywhere in the code base according to my `git grep`. We might want to remove the usage of `self.context` throughout the whole class.

Reply:
It would be a good idea to exclude everything under .git if it exists. Otherwise, the rationale is that ``git ls-files`` returns a "dumb" list of strings while ``os.walk`` provides a list of path-like objects, including whether it is a symlink or not so we can "resolve" the full path with ``os.path.realpath``.

The fact that a path is a symlink is important for identifying the right file.
For example, [plugins/modules/zypper.py](https://github.com/ansible-collections/community.general/blob/main/plugins/modules/zypper.py) is a symlink to [plugins/modules/packaging/os/zypper.py](https://github.com/ansible-collections/community.general/blob/main/plugins/modules/packaging/os/zypper.py).

Without this patch, the bot would identify it as as the symlink rather than the target.